aws/resource_aws_traffic_mirror_filter_rule_test.go

```diff
@@ -0,0 +1,249 @@
+			TrafficMirrorFilterIds: []*string{
+				aws.String(filterId),
+			},
+		})
+
+		if err != nil {
+			return err
+		}
+
+		if 0 == len(out.TrafficMirrorFilters) {
+			return fmt.Errorf("Traffic mirror filter %s not found", rs.Primary.ID)
+		}
+
+		filter := out.TrafficMirrorFilters[0]
+		var ruleList []*ec2.TrafficMirrorFilterRule
+		ruleList = append(ruleList, filter.IngressFilterRules...)
+		ruleList = append(ruleList, filter.EgressFilterRules...)
+
+		var exists bool
+		for _, rule := range ruleList {
+			if *rule.TrafficMirrorFilterRuleId == ruleId {
+				exists = true
+				break
+			}
+		}
+
+		if !exists {
+			return fmt.Errorf("Rule %s not found inside filter %s", ruleId, filterId)
+		}
+
+		return nil
+	}
+}
+
+func testAccTrafficMirrorFilterRuleConfig(dstCidr, srcCidr, action, dir string, num int) string {
+	return fmt.Sprintf(`
+resource "aws_traffic_mirror_filter" "filter" {
+}
+
+resource "aws_traffic_mirror_filter_rule" "rule" {
+	traffic_mirror_filter_id = "${aws_traffic_mirror_filter.filter.id}"
+	destination_cidr_block = "%s"
+	rule_action = "%s"
+	rule_number = %d
+	source_cidr_block = "%s"
+	traffic_direction = "%s"
+}
+`, dstCidr, action, num, srcCidr, dir)
+}
+
+func testAccTrafficMirrorFilterRuleConfigFull(dstCidr, srcCidr, action, dir, description string, ruleNum, srcPortFrom, srcPortTo, dstPortFrom, dstPortTo, protocol int) string {
+	return fmt.Sprintf(`
+resource "aws_traffic_mirror_filter" "filter" {
+}
+
+resource "aws_traffic_mirror_filter_rule" "rule" {
+	traffic_mirror_filter_id = "${aws_traffic_mirror_filter.filter.id}"
+	destination_cidr_block = "%s"
+	rule_action = "%s"
+	rule_number = %d
+	source_cidr_block = "%s"
+	traffic_direction = "%s"
+	description = "%s"
+	protocol = %d
+	source_port_range {
+		from_port = %d
+		to_port = %d
+	}
+	destination_port_range {
+		from_port = %d
+		to_port = %d
+	}
+}
+`, dstCidr, action, ruleNum, srcCidr, dir, description, protocol, srcPortFrom, srcPortTo, dstPortFrom, dstPortTo)
+}
+
+func testAccPreCheckAWSTrafficMirrorFilterRule(t *testing.T) {
+	conn := testAccProvider.Meta().(*AWSClient).ec2conn
+
+	_, err := conn.DescribeTrafficMirrorFilters(&ec2.DescribeTrafficMirrorFiltersInput{})
+
+	if testAccPreCheckSkipError(err) {
+		t.Skip("skipping traffic mirror filter rule acceprance test: ", err)
+	}
+
+	if err != nil {
+		t.Fatal("Unexpected PreCheck error: ", err)
+	}
+}
+
+func testAccCheckAwsTrafficMirrorFilterRuleDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).ec2conn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_traffic_mirror_filter_rule" {
+			continue
+		}
+
+		ruleId := rs.Primary.ID
+		filterId := rs.Primary.Attributes["traffic_mirror_filter_id"]
+
+		out, err := conn.DescribeTrafficMirrorFilters(&ec2.DescribeTrafficMirrorFiltersInput{
+			TrafficMirrorFilterIds: []*string{
+				aws.String(filterId),
+			},
+		})
+
+		if isAWSErr(err, "InvalidTrafficMirrorFilterId.NotFound", "") {
```

Comment:
If this is the error that is returned when an EC2 Traffic Mirror Filter cannot be found, we should include this in the filter and filter rule resource Read functions to trigger Terraform resource recreation:

```go
// Resource Read functions, potentially in addition to over nil/length checks from DescribeTrafficMirrorFilters call
if isAWSErr(err, "InvalidTrafficMirrorFilterId.NotFound", "") {
	log.Printf("[WARN] EC2 Traffic Mirror Filter (%s) not found, removing from state", d.Id())
	d.SetId("")
	return nil
}
```